walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,187 @@
+        Math.max(firstTimeRange.end(), secondTimeRange.end()),
+        false
+    );
+
+    return mergedTimeRange;
+  }
+
+  private List<TimeRange> getAvailableTimeRanges(List<TimeRange> unavailableTimeRanges, long duration) {
+    if (unavailableTimeRanges.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> availableTimeRanges = new ArrayList<>();
+
+    // Check for available time starting at the beginning of the day.
+    if (unavailableTimeRanges.get(0).start() >= duration) {
+      availableTimeRanges.add(TimeRange.fromStartEnd(
+          TimeRange.WHOLE_DAY.start(),
+          unavailableTimeRanges.get(0).start(),
+          false
+      ));
+    }
+    
+    int i;
+    for (i = 0; i < unavailableTimeRanges.size() - 1; i++) {
+        if(enoughTimeBetween(unavailableTimeRanges.get(i), unavailableTimeRanges.get(i + 1), duration)) {
+          availableTimeRanges.add(TimeRange.fromStartEnd(
+              unavailableTimeRanges.get(i).end(),
+              unavailableTimeRanges.get(i + 1).start(),
+              false
+          ));
+        }
+    }
+
+    // Check for available time at the end of the day.
+    int timeAfterLastMeeting = TimeRange.WHOLE_DAY.end() - unavailableTimeRanges.get(i).end();
+    if (timeAfterLastMeeting >= duration) {
+      availableTimeRanges.add(TimeRange.fromStartEnd(
+          unavailableTimeRanges.get(unavailableTimeRanges.size() - 1).end(),
+          TimeRange.WHOLE_DAY.end(),
+          false
+      ));
+    }
+
+    return availableTimeRanges;
+  }
+
+  private List<TimeRange> mergeAvailableTimeRanges(
+      List<TimeRange> mandatory, List<TimeRange> optional, long duration) {
+    List<TimeRange> combined = new ArrayList<>();
+    int i = 0;
+    int j = 0;
+    while(i < mandatory.size() && j < optional.size()) {
+      TimeRange mandatoryRange = mandatory.get(i);
+      TimeRange optionalRange = optional.get(j);
+
+      // Start by comparing the first two TimeRanges
+      if (mandatoryRange.overlaps(optionalRange)) {
+        int rangeStart = Math.max(mandatoryRange.start(), optionalRange.start());
+        int rangeEnd = Math.min(mandatoryRange.end(), optionalRange.end());
+      
+        if (rangeEnd - rangeStart >= duration) {
+          combined.add(TimeRange.fromStartEnd(
+              rangeStart,
+              rangeEnd,
+              false
+          ));
+        }
+      }
+      
+      // Move to the next index depending on which lists' current timeRange
+      // ends first since the timeRange which ends later can be merged with
+      // another or multiple other timeRanges.
+      if (mandatoryRange.end() >= optionalRange.end()) {
+          j++;
+      } else {
+          i++;
+      }
+    }
+
+    return combined;
+  }
+
+   private boolean enoughTimeBetween(TimeRange firstTime, TimeRange nextTime, long duration) {
```

Comment:
There seems to be an extra space in the beginning.

Reply:
Done.